x/ccv/provider/keeper/distribution.go

```diff
@@ -57,32 +68,165 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
 	return consumerRewardDenoms
 }
 
-// TransferRewardsToFeeCollector transfers all consumer rewards to the fee collector address
-func (k Keeper) TransferRewardsToFeeCollector(ctx sdk.Context) {
-	// 1. Get the denom whitelist from the store
-	denoms := k.GetAllConsumerRewardDenoms(ctx)
+// AllocateTokens performs reward and fee distribution to all validators based
+// on the Partial Set Security distribution specification.
+func (k Keeper) AllocateTokens(ctx sdk.Context, totalPreviousPower int64, bondedVotes []abci.VoteInfo) {
+	// return if there is no coins in the consumer rewards pool
+	if k.GetConsumerRewardsPool(ctx).IsZero() {
+		return
+	}
 
-	// 2. Iterate over the whitelist
-	for _, denom := range denoms {
-		// 3. For each denom, retrieve the balance from the consumer rewards pool
-		balance := k.bankKeeper.GetBalance(
+	// Iterate over all registered consumer chains
+	for _, consumer := range k.GetAllConsumerChains(ctx) {
+
+		// transfer the consumer rewards to the distribution module account
+		// note that the rewards transferred are only consumer whitelisted denoms
+		rewardsCollected, err := k.TransferConsumerRewardsToDistributionModule(ctx, consumer.ChainId)
+		if err != nil {
+			panic(err)
+		}
+
+		if rewardsCollected.IsZero() {
+			continue
+		}
+
+		rewardsCollectedDec := sdk.NewDecCoinsFromCoins(rewardsCollected...)
+
+		// temporary workaround to keep CanWithdrawInvariant happy
+		// general discussions here: https://github.com/cosmos/cosmos-sdk/issues/2906#issuecomment-441867634
+		feePool := k.distributionKeeper.GetFeePool(ctx)
+		if totalPreviousPower == 0 {
+			feePool.CommunityPool = feePool.CommunityPool.Add(rewardsCollectedDec...)
+			k.distributionKeeper.SetFeePool(ctx, feePool)
+			return
+		}
+
+		// Calculate the reward allocations
+		remaining := rewardsCollectedDec
+		communityTax := k.distributionKeeper.GetCommunityTax(ctx)
+		voteMultiplier := math.LegacyOneDec().Sub(communityTax)
+		feeMultiplier := rewardsCollectedDec.MulDecTruncate(voteMultiplier)
+
+		// allocate tokens to consumer validators
+		feeAllocated := k.AllocateTokensToConsumerValidators(
+			ctx,
+			consumer.ChainId,
+			totalPreviousPower,
+			bondedVotes,
+			feeMultiplier,
+		)
+		remaining = remaining.Sub(feeAllocated)
+
+		// allocate community funding
+		feePool.CommunityPool = feePool.CommunityPool.Add(remaining...)
+		k.distributionKeeper.SetFeePool(ctx, feePool)
+	}
+}
+
+// TODO: define which validators earned the tokens, i.e. already opted in for 1000 blocks
+// AllocateTokensToConsumerValidators allocates the consumer rewards pool to validator
+// according to their voting power
+func (k Keeper) AllocateTokensToConsumerValidators(
+	ctx sdk.Context,
+	chainID string,
+	totalPreviousPower int64,
+	bondedVotes []abci.VoteInfo,
+	tokens sdk.DecCoins,
+) sdk.DecCoins {
+	optedInVals := map[string]struct{}{}
```

Comment:
Could we just get rid of the whole "checking for opted-in" logic here? I know it's not the final form, but what's there right now is weird to me.
1) only opted in validators could even vote, so I think the whole check whether vals are opted in right now seems unnecessary
2) it probably shouldn't matter whether a val is opted in *right now* - e.g. this behaves weirdly when a validator opts out, but then rewards for a block come in where the val was still validating.

Reply:
Works for me! Will add a TODO instead.